components/script/dom/servoparser/async_html.rs

```diff
@@ -0,0 +1,561 @@
+            };
+
+            HtmlTokenizer::new(tb, tok_options)
+        } else {
+            HtmlTokenizer::new(TreeBuilder::new(sink, options), Default::default())
+        };
+
+        Tokenizer {
+            inner: inner,
+        }
+    }
+
+    pub fn feed(&mut self, input: &mut BufferQueue) -> Result<(), Root<HTMLScriptElement>> {
+        match self.inner.feed(input) {
+            TokenizerResult::Done => Ok(()),
+            TokenizerResult::Script(script) => {
+                let nodes = self.inner.sink.sink.nodes.borrow();
+                let script = nodes.get(&script.id).unwrap();
+                Err(Root::from_ref(script.downcast().unwrap()))
+            },
+        }
+    }
+
+    pub fn end(&mut self) {
+        self.inner.end();
+    }
+
+    pub fn url(&self) -> &ServoUrl {
+        &self.inner.sink.sink.base_url
+    }
+
+    pub fn set_plaintext_state(&mut self) {
+        self.inner.set_plaintext_state();
+    }
+}
+
+#[allow(unsafe_code)]
+unsafe impl JSTraceable for HtmlTokenizer<TreeBuilder<ParseNode, Sink>> {
+    unsafe fn trace(&self, trc: *mut JSTracer) {
+        struct Tracer(*mut JSTracer);
+        let tracer = Tracer(trc);
+
+        impl HtmlTracer for Tracer {
+            type Handle = ParseNode;
+            #[allow(unrooted_must_root)]
+            fn trace_handle(&self, node: &ParseNode) {
+                unsafe { node.trace(self.0); }
+            }
+        }
+
+        let tree_builder = &self.sink;
+        tree_builder.trace_handles(&tracer);
+        tree_builder.sink.trace(trc);
+    }
+}
+
+#[derive(JSTraceable, Clone, HeapSizeOf)]
+pub struct ParseNode {
+    id: usize,
+    qual_name: Option<QualName>,
+}
+
+enum ParserNodeOrText {
+    Node(usize),
+    Text(String),
+}
+
+#[derive(JSTraceable, HeapSizeOf)]
+struct ParseNodeData {
+    parent: Option<usize>,
+    target: Option<String>,
+    data: Option<String>,
+    contents: Option<ParseNode>,
+    is_integration_point: bool,
+}
+
+impl Default for ParseNodeData {
+    fn default() -> ParseNodeData {
+        ParseNodeData {
+            parent: None,
+            target: None,
+            data: None,
+            contents: None,
+            is_integration_point: false,
+        }
+    }
+}
+
+enum ParseOperation {
+    GetTemplateContents(usize, usize),
+    CreateElement(usize, QualName, Vec<Attribute>),
+    CreateComment(String, usize),
+    Insert(usize, Option<usize>, ParserNodeOrText),
+    AppendDoctypeToDocument(String, String, String),
+    AddAttrsIfMissing(usize, Vec<Attribute>),
+    RemoveFromParent(usize),
+    MarkScriptAlreadyStarted(usize),
+    ReparentChildren(usize, usize),
+    AssociateWithForm(usize, usize),
+    CreatePI(usize),
+    Pop(usize),
+}
+
+#[derive(JSTraceable, HeapSizeOf)]
+#[must_root]
+pub struct Sink {
+    base_url: ServoUrl,
+    document: JS<Document>,
+    current_line: u64,
+    script: MutNullableJS<HTMLScriptElement>,
+    parse_node_data: DOMRefCell<HashMap<usize, ParseNodeData>>,
```

Comment:
This is a tricky one, `parse_node_data` needs to be `DOMRefCell` because I want to use it mutably in the `same_tree` method, where I update the tree-related info based on the data I get from the updated `Node`s (`same tree` borrows self immutably).

The `nodes` field needs it because it is used in `process_operation`, which takes &self. If `process_operation` begins taking `&mut self`, it would cause problems with the TreeSink methods, as many methods already immutably borrow `parse_node_data` there, when they call `process_operation`. Therefore, the `process_operation` must borrow immutably too, to satisfy the borrow checker.

However, I might be able to make `nodes` a normal Hashmap in the next PR, because it will exist in main thread altogether, and therefore there will be no conflicts with `Sink` then.